### PR TITLE
Core/MMAPs: Adjust walkable climb and fix a lot of mmap raycast issues

### PR DESCRIPTION
**Pull request by jackpoz**
**Changes proposed:**

-  Adjust walkable climb as recast using walkableClimb also to find the poly from a position, giving priority to polys that are below the position but closer than walkableClimb.
- Fix a lot of mmap raycast issues
- Note that some of the fixes require to re-extract mmaps, even if the previous ones are still compatible

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:** Closes #  (insert issue tracker number)

**Tests performed:** (Does it build, tested in-game, etc.)
- .go xyz 3328.058838 -3476.440918 287.074432 533 1.060287
- .go xyz 3175.989990 -3469.872070 287.049133 533 1.698029
- .go xyz 1438.516113 -4422.130371 25.462406 1 3.292355
- .go xyz -8461.000000 1012.000000 59.473511 0 5.808

**Known issues and TODO list:** (add/remove lines as needed)

- .go xyz 1027.279053 -4941.14502 12.63004 1 3.926178 (but works for me)
- .go xyz 1293.80725 -4403.5991 26.312632 1 0.153907 (but works for me)
- .go xyz 1268.545 -4435.84 26.637 1 3.740760 (but works for me)

### CODE REVIEW

**Comment by Faq** [comment]
Any exact cases where this can be tested?

**Comment by Ovahlord** [comment]
Naxxramas, Anub'Rekhan's spawn position. Try to blink into the left wall behind him. With the current checks you should end up not moving at all because in the current state, the path will go out of bounds, resulting in the vmap checks blocking the path to not fall out of the map

**Comment by jackpoz** [comment]
To check: why the hitpoint sometimes has no height in the mmap mesh

**Comment by jackpoz** [comment]
Found the reason: floating point precision issues

**Comment by jackpoz** [comment]
@Ovahlord do we even need anymore a full path when using raycast, even in case we hit no wall ? https://github.com/TrinityCore/TrinityCore/commit/4a197ba22af4eed01be632ea2dd7d103a963f043 tried to create a point path but failed retrieving the Z from mmaps, causing the issue you saw on Stormwind bridge with points under the bridge.

One could call getPolyHeight() for each point trying to guess which poly is the current one (or iterating them, even if they might not be in order) or just drop the whole "build a point path for raycast" thing.

I checked the sources and I didn't find any other usage of mmaps raycast other than the one we do.

**Comment by jackpoz** [comment]
- [ ] To check: cases where raycast doesn't hit anything but height of endPos cannot be retrieved

**Comment by jackpoz** [comment]
This PR is ready for testing

**Comment by jackpoz** [comment]
PR hosted at https://tcubuntu.northeurope.cloudapp.azure.com/test/

**Comment by jackpoz** [comment]
To fix:
- .go xyz 3175.989990 -3469.872070 287.049133 533 1.698029 case where endPoly is invalid (but raycast doesn't need endPoly

**Comment by jackpoz** [comment]
Fixed

File: src/server/game/Entities/Object/Object.cpp [hunk 1]
```diff
@@ -3285,7 +3286,9 @@ void WorldObject::MovePositionToFirstCollision(Position &pos, float dist, float
 
     // check static LOS
     float halfHeight = GetCollisionHeight() * 0.5f;
-    bool col = VMAP::VMapFactory::createOrGetVMapManager()->getObjectHitPos(GetMapId(),
+    bool col;
+    /*
```
Comment (jackpoz) [inline src/server/game/Entities/Object/Object.cpp:3290]: this can stay commented for a while in case we have to restore it

**Comment by Killyana** [comment]
Unable to blink in:  1027.279053, -4941.14502, 12.63004, 3.926178 even if you can walk on this path.

**Comment by Jildor** [comment]
LK encounter platform
```
.go xyz 493.110474 -2120.135742 840.857117 631
```
use blink spellid: 1953
without this PR you don't move with blink
with this PR seems works fine
(same in Trial of the crusader and Malygos platforms)

**Comment by Jildor** [comment]
.go xyz -4884.618164 -1956.716309 89.984360 1
get fear (.aura 5782)
without PR you go to the hell
wirh PR seems more accurate (you don't move alot, but you can't fall easy xD)

**Comment by jackpoz** [comment]
@Killyana .go xyz 1027.279053 -4941.14502 12.63004 1 3.926178 worked for me on the test realm with a Human mage

**Comment by Jildor** [comment]
Maybe add a new version of mmaps to ensure people re-extract mmaps can be better

**Comment by jackpoz** [comment]
@Ovahlord did you have anything else to change before we can merge this PR ? you can test it at https://tcubuntu.northeurope.cloudapp.azure.com/test/

**Comment by Jildor** [comment]
seems with this PR fixes this issue: #24628

**Comment by jackpoz** [comment]
nice

**Pull request merged**